File: python/docsearch/fetch.py
"""Fetching, with the manners and the guards a crawler owes a stranger's site.

Everything here answers to the cache: a response that arrives is stored, and a
response that has not changed is served from storage without transferring a
body. Nothing else in the ingest path talks to the network.
"""

from __future__ import annotations

import hashlib
import sqlite3
import time
import urllib.robotparser
from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from urllib.parse import urljoin, urlsplit, urlunsplit

import httpx

from . import fetchcache, urlguard
from .urlguard import BlockedURLError

__all__ = ["FetchError", "Fetched", "Fetcher", "normalize"]

#: Identifies the crawler and where to complain about it.
USER_AGENT = "docsearch/0.1 (+https://github.com/bamsammich/docsearch)"

#: Seconds between requests to one host. A documentation site is someone
#: else's server, and a crawl of a few hundred pages has no reason to hurry.
DEFAULT_INTERVAL = 0.5

#: A redirect chain longer than this is a loop or a trap, not navigation.
MAX_REDIRECTS = 5

#: Backstop against a seed that turns out to address far more than a manual.
#: The crawler applies the real budget; this is the floor under a bug.
MAX_FETCHES = 5000

DEFAULT_TIMEOUT = 20.0

#: Status codes that carry a Location rather than a body.
_REDIRECTS = frozenset({301, 302, 303, 307, 308})


class FetchError(Exception):
    """A URL could not be fetched. Carries a reason for the operator."""


@dataclass(slots=True, frozen=True)
class Fetched:
    url: str
    final_url: str
    status: int
    body: bytes
    content_type: str | None
    from_cache: bool


def _remove_dot_segments(path: str) -> str:
    """RFC 3986 5.2.4.

    Written out rather than delegated to ``posixpath.normpath``, which strips a
    trailing slash. That slash is not decoration: ``/docs/`` and ``/docs`` are
    different resources, and on at least one real generator one of them is the
    page and the other is a 404.
    """
    out: list[str] = []
    for seg in path.split("/"):
        if seg == ".":
            continue
        if seg == "..":
            if len(out) > 1:
                out.pop()
            continue
        out.append(seg)
    joined = "/".join(out)
    # split() drops no information, so a path that ended in "." or ".." must
    # regain the slash those segments stood in for.
    if path.endswith(("/.", "/..")) and not joined.endswith("/"):
        joined += "/"
    return joined


def normalize(raw: str) -> str:
    """One spelling per resource, so the frontier and the cache agree.

    Lowercases scheme and host, drops the default port and the fragment, and
    resolves dot segments. Deliberately leaves the trailing slash, the query
    and its ordering alone: all three can change which resource is addressed.
    """
    parts = urlsplit(raw)
    scheme = parts.scheme.lower()
    try:
        host = (parts.hostname or "").lower()
        port = parts.port
    except ValueError as exc:
        raise FetchError(f"unparseable URL: {raw}") from exc
    if (scheme == "http" and port == 80) or (scheme == "https" and port == 443):
        port = None

    netloc = f"[{host}]" if ":" in host else host
    if port is not None:
        netloc = f"{netloc}:{port}"

    path = _remove_dot_segments(parts.path) or "/"
    return urlunsplit((scheme, netloc, path, parts.query, ""))


def _host_of(url: str) -> str:
    return (urlsplit(url).hostname or "").lower()


def _origin_of(url: str) -> str:
    parts = urlsplit(url)
    return f"{parts.scheme}://{parts.netloc}"


class Fetcher:
    """Fetches URLs through the cache, the guard, robots and a rate limit.

    ``guard`` is injectable so tests can exercise the transport against a local
    server; the default refuses loopback, which is exactly what a test server
    is. That the default blocks it is covered by the guard's own tests.
    """

    def __init__(
        self,
        cache: sqlite3.Connection,
        *,
        client: httpx.Client | None = None,
        guard: Callable[[str], object] = urlguard.check,
        addr_guard: Callable[[str], bool] = urlguard.addr_allowed,
        user_agent: str = USER_AGENT,
        interval: float = DEFAULT_INTERVAL,
        max_redirects: int = MAX_REDIRECTS,
        max_fetches: int = MAX_FETCHES,
        obey_robots: bool = True,
    ) -> None:
        self.cache = cache
        self.guard = guard
        self.addr_guard = addr_guard
        self.user_agent = user_agent
        self.interval = interval
        self.max_redirects = max_redirects
        self.max_fetches = max_fetches
        self.obey_robots = obey_robots
        self._client = client or httpx.Client(
            follow_redirects=False,
            timeout=DEFAULT_TIMEOUT,
            headers={"User-Agent": user_agent},
        )
        self._owns_client = client is None
        self._last_request: dict[str, float] = {}
        self._robots: dict[str, urllib.robotparser.RobotFileParser] = {}
        self._fetches = 0

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> Fetcher:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    # -- politeness --------------------------------------------------------
    def _wait(self, host: str) -> None:
        last = self._last_request.get(host)
        if last is not None:
            remaining = self.interval - (time.monotonic() - last)
            if remaining > 0:
                time.sleep(remaining)
        self._last_request[host] = time.monotonic()

    def _robots_for(self, url: str) -> urllib.robotparser.RobotFileParser:
        host = _host_of(url)
        cached = self._robots.get(host)
        if cached is not None:
            return cached

        parser = urllib.robotparser.RobotFileParser()
        body = fetchcache.get_robots(self.cache, host)
        if body is None:
            body = ""
            try:
                # Not routed through fetch(): asking robots.txt whether
                # robots.txt may be fetched does not terminate. It is still
                # streamed and peer-checked, because this was the one request
                # the fetcher made that nothing verified had arrived from the
                # address the guard approved.
                robots_url = urljoin(_origin_of(url), "/robots.txt")
                self.guard(robots_url)
                self._wait(host)
                with self._client.stream("GET", robots_url) as res:
                    self._check_peer(res)
                    if res.status_code == 200:
                        body = res.read().decode("utf-8", errors="replace")
            except (BlockedURLError, FetchError, httpx.HTTPError):
                # A host that will not serve robots.txt has not disallowed
                # anything. Absent is permission, per the standard.
                body = ""
            fetchcache.put_robots(self.cache, host, body)
        parser.parse(body.splitlines())
        self._robots[host] = parser
        return parser

    def allowed_by_robots(self, url: str) -> bool:
        if not self.obey_robots:
            return True
        return self._robots_for(url).can_fetch(self.user_agent, url)

    def _check_peer(self, res: httpx.Response) -> None:
        """Refuse a response that arrived from an address the guard rejects.

        The name was validated before the request, but a resolver is free to
        answer differently the second time. Checked here, while the connection
        is open and before any body is read, so a rebound name cannot get its
        content into the index.

        A backstop rather than the boundary: the request has already gone out.
        Pinning the connection to the address that was validated would close
        the window entirely, and needs the transport to accept a resolved
        address while still presenting the original name for TLS.

        Fails closed when the peer cannot be determined. Returning quietly
        meant a transport that exposes no address disabled the check silently,
        which is the shape of gap that survives review: nothing errors, the
        body is read, and the only evidence is an absence. Every transport this
        fetcher uses does expose one, so refusing costs nothing and stops a
        future one from turning the check off by accident.
        """
        stream = res.extensions.get("network_stream")
        if stream is None:
            raise FetchError(
                "the transport exposed no connection to inspect, so the peer address "
                "could not be confirmed against the guard"
            )
        info = stream.get_extra_info("server_addr")
        if not info:
            raise FetchError(
                "the connection reported no peer address, so it could not be confirmed "
                "against the guard"
            )
        host = info[0] if isinstance(info, tuple) else str(info)
        if not self.addr_guard(str(host)):
            raise FetchError("response arrived from an address that is not permitted")

    # -- fetching ----------------------------------------------------------
    def fetch(self, raw: str, *, revalidate: bool = True) -> Fetched:
        """Fetch one URL, serving an unchanged body from the cache.

        ``revalidate=False`` returns any stored copy without asking the server,
        which is what re-chunking an already-crawled site wants.
        """
        url = normalize(raw)
        cached = fetchcache.get(self.cache, url)
        if cached is not None and not revalidate:
            return _from_cache(cached)

        if not self.allowed_by_robots(url):
            raise FetchError(f"robots.txt disallows {url}")
        if self._fetches >= self.max_fetches:
            raise FetchError(f"fetch budget of {self.max_fetches} exhausted")

        headers: dict[str, str] = {}
        if cached is not None:
            # Conditional: a site that answers 304 costs a round trip and no
            # body, which is what makes refreshing a large site cheap.
            if cached.etag:
                headers["If-None-Match"] = cached.etag
            if cached.last_modified:
                headers["If-Modified-Since"] = cached.last_modified

        current = url
        for _ in range(self.max_redirects + 1):
            self.guard(current)  # every hop, not just the seed
            self._wait(_host_of(current))
            self._fetches += 1
            try:
                # Streamed so the peer can be checked while the connection is
                # still open, and before a byte of body is read. Once the
                # response is buffered the socket is back in the pool and its
                # address is no longer available to ask.
                with self._client.stream("GET", current, headers=headers) as res:
                    self._check_peer(res)
                    status = res.status_code
                    location = res.headers.get("location")
                    content_type = res.headers.get("content-type")
                    etag = res.headers.get("etag")
                    last_modified = res.headers.get("last-modified")
                    body = b"" if status in _REDIRECTS or status == 304 else res.read()
            except httpx.HTTPError as exc:
                raise FetchError(f"{current}: {type(exc).__name__}") from exc

            if status in _REDIRECTS:
                if not location:
                    raise FetchError(f"{current}: redirect without a location")
                current = normalize(urljoin(current, location))
                continue

            if status == 304 and cached is not None:
                fetchcache.touch(self.cache, url)
                return _from_cache(cached)

            entry = fetchcache.CachedResponse(
                url=url,
                final_url=current,
                status=status,
                content_type=content_type,
                etag=etag,
                last_modified=last_modified,
                body=body,
                sha256=hashlib.sha256(body).hexdigest(),
                fetched_at=datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
            )
            fetchcache.put(self.cache, entry)
            return Fetched(
                url=url,
                final_url=current,
                status=status,
                body=body,
                content_type=content_type,
                from_cache=False,
            )

        raise FetchError(f"{url}: more than {self.max_redirects} redirects")


def _from_cache(entry: fetchcache.CachedResponse) -> Fetched:
    return Fetched(
        url=entry.url,
        final_url=entry.final_url,
        status=entry.status,
        body=entry.body,
        content_type=entry.content_type,
        from_cache=True,
    )

File: internal/site/fetch/fetch.go
package fetch

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/netip"
	"net/url"
	"time"

	"github.com/jimsmart/grobotstxt"

	"github.com/bamsammich/docsearch/internal/urlguard"
)

// UserAgent identifies the crawler and where to complain about it.
const UserAgent = "docsearch/0.1 (+https://github.com/bamsammich/docsearch)"

const (
	// DefaultInterval is the wait between requests to one host. A
	// documentation site is someone else's server, and a crawl of a few
	// hundred pages has no reason to hurry.
	DefaultInterval = 500 * time.Millisecond
	// DefaultMaxRedirects is where a redirect chain becomes a loop or a
	// trap rather than navigation.
	DefaultMaxRedirects = 5
	// DefaultMaxFetches is a backstop against a seed that addresses far
	// more than a manual. The crawler applies the real budget; this is the
	// floor under a bug.
	DefaultMaxFetches = 5000
	// DefaultTimeout bounds one request.
	DefaultTimeout = 20 * time.Second
)

// ErrFetch is returned for a URL that could not be fetched, with a reason
// for the operator.
var ErrFetch = errors.New("fetch failed")

// Fetched is one response, from the network or from the cache.
type Fetched struct {
	URL         string
	FinalURL    string
	ContentType string
	Body        []byte
	Status      int
	FromCache   bool
}

// Guard decides whether a URL may be fetched, and at which addresses.
type Guard func(ctx context.Context, raw string) (*urlguard.Target, error)

// Options configure a Fetcher. The zero value of each field takes the
// default beside it.
type Options struct {
	// Guard checks each URL. A nil guard uses urlguard.Check, which refuses
	// loopback and every other address a crawler has no business reaching;
	// a test server is exactly that, so tests supply their own.
	Guard Guard
	// Resolver looks up hosts for the default guard. A nil resolver uses
	// the system's.
	Resolver urlguard.Resolver
	// Transport dials and speaks HTTP. A nil transport dials only the
	// addresses the guard approved for the request's host.
	Transport http.RoundTripper
	UserAgent string
	// Interval is the wait between requests to one host.
	Interval     time.Duration
	Timeout      time.Duration
	MaxRedirects int
	MaxFetches   int
	// IgnoreRobots fetches without asking robots.txt. Reserved for a
	// caller fetching its own site.
	IgnoreRobots bool
}

// Fetcher fetches URLs through the cache, the guard, robots.txt and a rate
// limit. It is safe for one goroutine; the crawler fetches in order.
type Fetcher struct {
	cache        Cache
	client       *http.Client
	guard        Guard
	robots       map[string]string
	lastRequest  map[string]time.Time
	userAgent    string
	interval     time.Duration
	maxRedirects int
	maxFetches   int
	fetches      int
	ignoreRobots bool
}

// New returns a Fetcher storing into cache.
func New(cache Cache, opts Options) *Fetcher {
	f := &Fetcher{
		cache:        cache,
		guard:        opts.Guard,
		robots:       map[string]string{},
		lastRequest:  map[string]time.Time{},
		userAgent:    or(opts.UserAgent, UserAgent),
		interval:     orDuration(opts.Interval, DefaultInterval),
		maxRedirects: orInt(opts.MaxRedirects, DefaultMaxRedirects),
		maxFetches:   orInt(opts.MaxFetches, DefaultMaxFetches),
		ignoreRobots: opts.IgnoreRobots,
	}
	if f.guard == nil {
		resolver := opts.Resolver
		if resolver == nil {
			resolver = net.DefaultResolver
		}
		f.guard = func(ctx context.Context, raw string) (*urlguard.Target, error) {
			return urlguard.Check(ctx, raw, resolver)
		}
	}
	transport := opts.Transport
	if transport == nil {
		transport = pinnedTransport()
	}
	f.client = &http.Client{
		Transport: transport,
		Timeout:   orDuration(opts.Timeout, DefaultTimeout),
		// Redirects are followed by hand: every hop goes through the guard.
		CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse },
	}
	return f
}

// Fetch retrieves one URL, serving an unchanged body from the cache.
// revalidate false returns any stored copy without asking the server, which
// is what re-chunking an already-crawled site wants.
func (f *Fetcher) Fetch(ctx context.Context, raw string, revalidate bool) (*Fetched, error) {
	target, err := Normalize(raw)
	if err != nil {
		return nil, err
	}
	cached, hit, err := f.cache.Get(ctx, target)
	if err != nil {
		return nil, err
	}
	if hit && !revalidate {
		return fromCache(cached), nil
	}
	allowed, err := f.allowedByRobots(ctx, target)
	if err != nil {
		return nil, err
	}
	if !allowed {
		return nil, fmt.Errorf("%w: robots.txt disallows %s", ErrFetch, target)
	}
	return f.follow(ctx, target, cached, hit)
}

// follow requests the URL and each redirect it answers with, up to the
// limit.
func (f *Fetcher) follow(
	ctx context.Context,
	target string,
	cached *Response,
	hit bool,
) (*Fetched, error) {
	current := target
	for range f.maxRedirects + 1 {
		res, err := f.hop(ctx, current, conditional(cached, hit))
		if err != nil {
			return nil, err
		}
		if !isRedirect(res.Status) {
			return f.settle(ctx, target, current, res, cached, hit)
		}
		if current, err = redirectTo(current, res.Location); err != nil {
			return nil, err
		}
	}
	return nil, fmt.Errorf("%w: %s: more than %d redirects", ErrFetch, target, f.maxRedirects)
}

// hop requests one URL of a redirect chain, within the budget.
func (f *Fetcher) hop(
	ctx context.Context,
	raw string,
	headers map[string]string,
) (*response, error) {
	if f.fetches >= f.maxFetches {
		return nil, fmt.Errorf("%w: fetch budget of %d exhausted", ErrFetch, f.maxFetches)
	}
	return f.request(ctx, raw, headers)
}

// settle turns the response that ended the redirect chain into a result: a
// 304 confirms the stored copy, anything else replaces it.
func (f *Fetcher) settle(ctx context.Context, target, final string, res *response,
	cached *Response, hit bool,
) (*Fetched, error) {
	if res.Status == http.StatusNotModified && hit {
		if err := f.cache.Touch(ctx, target); err != nil {
			return nil, err
		}
		return fromCache(cached), nil
	}
	return f.store(ctx, target, final, res)
}

// store caches a response and returns it.
func (f *Fetcher) store(
	ctx context.Context,
	target, final string,
	res *response,
) (*Fetched, error) {
	sum := sha256.Sum256(res.Body)
	entry := &Response{
		URL: target, FinalURL: final, Status: res.Status, ContentType: res.ContentType,
		ETag: res.ETag, LastModified: res.LastModified, Body: res.Body,
		SHA256: hex.EncodeToString(sum[:]), FetchedAt: now(),
	}
	if err := f.cache.Put(ctx, entry); err != nil {
		return nil, err
	}
	return &Fetched{
		URL: target, FinalURL: final, Status: res.Status,
		Body: res.Body, ContentType: res.ContentType,
	}, nil
}

// response is one HTTP response, read.
type response struct {
	ContentType  string
	ETag         string
	LastModified string
	Location     string
	Body         []byte
	Status       int
}

// request checks the URL with the guard, waits out the host's interval, and
// reads the response.
func (f *Fetcher) request(
	ctx context.Context,
	raw string,
	headers map[string]string,
) (*response, error) {
	checked, err := f.guard(ctx, raw)
	if err != nil {
		return nil, fmt.Errorf("%w: %s: %w", ErrFetch, raw, err)
	}
	f.wait(checked.URL.Hostname())
	f.fetches++
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, raw, nil)
	if err != nil {
		return nil, fmt.Errorf("%w: %s: %w", ErrFetch, raw, err)
	}
	req.Header.Set("User-Agent", f.userAgent)
	for name, value := range headers {
		req.Header.Set(name, value)
	}
	return f.do(withAllowedAddrs(req, checked.Addrs))
}

func (f *Fetcher) do(req *http.Request) (*response, error) {
	res, err := f.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %s: %w", ErrFetch, req.URL, err)
	}
	defer res.Body.Close()
	out := &response{
		Status:       res.StatusCode,
		ContentType:  res.Header.Get("Content-Type"),
		ETag:         res.Header.Get("ETag"),
		LastModified: res.Header.Get("Last-Modified"),
		Location:     res.Header.Get("Location"),
	}
	if isRedirect(out.Status) || out.Status == http.StatusNotModified {
		// Neither carries a body worth reading.
		return out, nil
	}
	body, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, fmt.Errorf("%w: %s: %w", ErrFetch, req.URL, err)
	}
	out.Body = body
	return out, nil
}

// wait keeps one request at a time to a host, spaced by the interval.
func (f *Fetcher) wait(host string) {
	if last, ok := f.lastRequest[host]; ok {
		if remaining := f.interval - time.Since(last); remaining > 0 {
			time.Sleep(remaining)
		}
	}
	f.lastRequest[host] = time.Now()
}

// allowedByRobots reports whether the host's robots.txt permits the URL.
func (f *Fetcher) allowedByRobots(ctx context.Context, raw string) (bool, error) {
	if f.ignoreRobots {
		return true, nil
	}
	u, err := url.Parse(raw)
	if err != nil {
		return false, fmt.Errorf("%w: %s: %w", ErrFetch, raw, err)
	}
	body, err := f.robotsFor(ctx, u)
	if err != nil {
		return false, err
	}
	return grobotstxt.AgentAllowed(body, f.userAgent, raw), nil
}

// robotsFor reads a host's robots.txt, from memory, then the cache, then
// the host. A host that will not serve one has disallowed nothing: absent
// is permission, per the standard.
func (f *Fetcher) robotsFor(ctx context.Context, u *url.URL) (string, error) {
	host := u.Hostname()
	if body, ok := f.robots[host]; ok {
		return body, nil
	}
	body, ok, err := f.cache.Robots(ctx, host)
	if err != nil {
		return "", err
	}
	if !ok {
		// Not routed through Fetch: asking robots.txt whether robots.txt may
		// be fetched does not terminate.
		body = f.fetchRobots(ctx, u.Scheme+"://"+u.Host+"/robots.txt")
		if err := f.cache.PutRobots(ctx, host, body); err != nil {
			return "", err
		}
	}
	f.robots[host] = body
	return body, nil
}

// fetchRobots reads a host's robots.txt, or "" when it does not serve one.
func (f *Fetcher) fetchRobots(ctx context.Context, raw string) string {
	res, err := f.request(ctx, raw, nil)
	if err != nil || res.Status != http.StatusOK {
		return ""
	}
	return string(res.Body)
}

// conditional asks the server for the body only if it changed, which is
// what makes refreshing a large site cheap.
func conditional(cached *Response, hit bool) map[string]string {
	headers := map[string]string{}
	if !hit {
		return headers
	}
	if cached.ETag != "" {
		headers["If-None-Match"] = cached.ETag
	}
	if cached.LastModified != "" {
		headers["If-Modified-Since"] = cached.LastModified
	}
	return headers
}

// redirectTo is the normalized target of a redirect.
func redirectTo(from, location string) (string, error) {
	if location == "" {
		return "", fmt.Errorf("%w: %s: redirect without a location", ErrFetch, from)
	}
	base, err := url.Parse(from)
	if err != nil {
		return "", fmt.Errorf("%w: %s: %w", ErrFetch, from, err)
	}
	next, err := base.Parse(location)
	if err != nil {
		return "", fmt.Errorf("%w: %s: unresolvable redirect to %s", ErrFetch, from, location)
	}
	return Normalize(next.String())
}

func isRedirect(status int) bool {
	switch status {
	case http.StatusMovedPermanently, http.StatusFound, http.StatusSeeOther,
		http.StatusTemporaryRedirect, http.StatusPermanentRedirect:
		return true
	}
	return false
}

func fromCache(r *Response) *Fetched {
	return &Fetched{
		URL: r.URL, FinalURL: r.FinalURL, Status: r.Status,
		Body: r.Body, ContentType: r.ContentType, FromCache: true,
	}
}

// allowedAddrsKey carries the addresses the guard approved for a request to
// the dialler, so the connection cannot go anywhere else.
type allowedAddrsKey struct{}

func withAllowedAddrs(req *http.Request, addrs []netip.Addr) *http.Request {
	return req.WithContext(context.WithValue(req.Context(), allowedAddrsKey{}, addrs))
}

// pinnedTransport dials only the addresses the guard approved for this
// request. The name was validated before the request; a resolver is free to
// answer differently the second time, and dialling the approved address
// closes that window rather than narrowing it. TLS still presents the name
// from the URL, so certificates verify as usual.
func pinnedTransport() http.RoundTripper {
	dialer := &net.Dialer{Timeout: 10 * time.Second, KeepAlive: 30 * time.Second}
	transport := http.DefaultTransport.(*http.Transport).Clone()
	transport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
		addrs, ok := ctx.Value(allowedAddrsKey{}).([]netip.Addr)
		if !ok || len(addrs) == 0 {
			return nil, fmt.Errorf("%w: no address was approved for %s", ErrFetch, addr)
		}
		return dialApproved(ctx, dialer, network, addr, addrs)
	}
	return transport
}

// dialApproved tries each approved address in turn, at the port the
// request named.
func dialApproved(ctx context.Context, dialer *net.Dialer, network, addr string,
	addrs []netip.Addr,
) (net.Conn, error) {
	_, port, err := net.SplitHostPort(addr)
	if err != nil {
		return nil, fmt.Errorf("%w: %s: %w", ErrFetch, addr, err)
	}
	var last error
	for _, a := range addrs {
		conn, dialErr := dialer.DialContext(ctx, network, net.JoinHostPort(a.String(), port))
		if dialErr == nil {
			return conn, nil
		}
		last = dialErr
	}
	return nil, fmt.Errorf("%w: %s: %w", ErrFetch, addr, last)
}

// or and its siblings take the default when a field was left zero.
func or(v, fallback string) string {
	if v == "" {
		return fallback
	}
	return v
}

func orInt(v, fallback int) int {
	if v == 0 {
		return fallback
	}
	return v
}

func orDuration(v, fallback time.Duration) time.Duration {
	if v == 0 {
		return fallback
	}
	return v
}
